---
title: "HW1"
author: "SES"
date: "2025-09-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/projects/eco_genomics_2025/population_genomics")
```

1. Import our libraries for plotting 
```{r}
#load plotting libraries 
library(tidyverse)
library(ggplot2)
library(dplyr)
library(ggVennDiagram)
library(viridisLite)
library(ggpubr)
library(qqman)
```
2.Using Nucleotide diversity to find selection signatures
```{r} 
setwd("myresults/ANGSD/diversity")

theta <- read.table("_ALL_win50000_step50000.thetas", sep = "\t", header = T)

head(theta)
#Scaling Watterson's theta by nSites
theta$tWsite = theta$tW/theta$nSites 

#Scaling Pi by nSites
theta$tPsite = theta$tP/theta$nSites 

#check!
head(theta)
summary(theta)

#rearrange theta table 
theta2 = theta %>% 
  arrange(Chr, WinCenter) %>%
  filter(Chr > 100)

#subsetting rows that are evidence of selective sweep
theta2_filt <- theta2[which(theta2$Tajima < -1.5 | theta2$tPsite < 0.001),]

# Create labels for plotting (Where is the significance coming from?)
theta2_filt$sig_status <- ifelse((theta2_filt$Tajima<-1.5) & (theta2_filt$tPsite<cutoff), "Both",
                       ifelse(theta2_filt$Tajima<-1.5, "Tajima",
                       ifelse(theta2_filt$tPsite<0.001, "Pi", "None")))

summary(theta2_filt)
sum(theta2_filt$nSites)

```

3. Using PCA to find selection signatures 
```{r}
setwd("myresults/ANGSD/PCA_ADMIX")

#making the PCA with just the Red Spruce 

allRScov <- as.matrix(read.table("allRS_poly_K3.cov"))

allRSPCA <- eigen(allRScov)

var <- round(allRSPCA$values/sum(allRSPCA$values),3)

var[1:3]

#Bring in the bam list files and extract sample name information. Line 40 is the data frame with the list of names. The last number is the number of samples (1-95)
names <- read.table("allRS_bam.list")
names <- unlist(strsplit(basename(as.character(names[,1])), split = ".sorted.rmdup.bam"))
split = strsplit(names, "_")
pops <- data.frame(names[1:95], do.call(rbind, split[1:95]))
names(pops) = c("Ind", "Pop", "Row", "Col")

s <- read.table("allRS_poly_K3.selection")

# convert test statistic to p-value
pval_PC1 <- as.data.frame(1-pchisq(s[,1],1))
names(pval_PC1) = "p_PC1"

pval_PC2 <- as.data.frame(1-pchisq(s[,2],1))
names(pval_PC2) = "p_PC2"

#read in the sites: data on the snp positions themselves. Pairs the site information with sites that were selected for analysis in the outlier test.

PCAngsd_sites <- read.table("allRS_poly_K3.sites")
MAF_all_sites <- read.table("/gpfs1/cl/ecogen/pbio6800/PopulationGenomics/ANGSD/allRS_poly.mafs.gz", header=T)

pos_filtered <- cbind(MAF_all_sites,PCAngsd_sites)[which(PCAngsd_sites==1),]

dim(pos_filtered)

# Combine the locus info with the PC outlier p-values:

loci_pvals = cbind(pos_filtered, pval_PC1, pval_PC2)


# get all the outliers with p-values below some cutoff
cutoff=1e-3   # equals p<0.001

#table of selection outliers for each axis including the contig, snp position, and p_values 
outliers_PC1 <- loci_pvals %>%
                    filter(p_PC1<cutoff) %>%
                    select(chromo,position,p_PC1,p_PC2)
                    
outliers_PC2 <- loci_pvals %>%
                    filter(p_PC2<cutoff) %>%
                    select(chromo,position,p_PC1,p_PC2)


# how many outlier loci < the cutoff?
dim(outliers_PC1)[1]

dim(outliers_PC2)[1]

# how many unique contigs harbor outlier loci?
outlier_contigs_PC1 = unique(outliers_PC1$chromo)

length(outlier_contigs_PC1)

outlier_contigs_PC2 = unique(outliers_PC2$chromo)

length(outlier_contigs_PC2)
```
4. comparing significant values of the PCA outlier contigs with Tajima's D and Pi 
```{r}
#Checking colnames of tajima'sD and Pi (theta2 table) selection signatures
colnames(theta2_filt)

#need to rename Chromo to Chr in the PCA outliers table to be able to compare the 3 tables 
#PC1
#outliers_PC1 <- rename(outliers_PC1, Chr=chromo)
#PC2
#outliers_PC2 <- rename(outliers_PC2, Chr=chromo)

##OVERLAP OF CONTIGS WITH SELECTION SIGNTURES (Venn Diagram)
venn_data <- list( 
  "TajimaD and Pi" = theta2_filt$Chr, 
  "PC1 outliers" = outlier_contigs_PC1, 
  "PC2 outliers" = outlier_contigs_PC2)

ggVennDiagram(venn_data, label_alpha = 0.5) +
  scale_fill_distiller(palette = "GnBu") +
  theme(
    legend.position = "right",
    venn.setLabel = element_text(size = 20, color = "black"),
    plot.title = element_text(hjust = 0.5, size = 28, face = "bold")
  )

#ggsave("~/projects/eco_genomics_2025/population_genomics/mydocs/venn.png", plot = venn, width = 6, height = 4, dpi = 300)


```

5. Signatures of Selection at the SNP level 
```{r}
#SNPS with Selection signatures

#combining the PC1 and PC2 outliers tables, getting rid of overlaps  
all_snps <- unique(rbind(outliers_PC1, outliers_PC2))

# Create labels for plotting
all_snps$SNP_status <- ifelse((all_snps$p_PC1<cutoff) & (all_snps$p_PC2<cutoff), "Both",
                       ifelse(all_snps$p_PC1<cutoff, "PC1_only",
                       ifelse(all_snps$p_PC2<cutoff, "PC2_only", "None")))

# Calculate -log10(p-value) from either axis (choose one for y-axis)
all_snps$logp <- -log10(pmin(all_snps$p_PC1, all_snps$p_PC2, na.rm = TRUE))

#-------Making Chr and SNP positions into Genomic Positions

#making sure the data frame and scaffolds are ordered properly 
all_snpsOrdered <- all_snps %>% arrange(Chr, position)
all_snpsOrdered$Chr <- factor(all_snpsOrdered$Chr, levels = unique(all_snpsOrdered$Chr))

# Step 1: Get scaffold offsets
scaffold_offsets <- all_snpsOrdered %>%
  group_by(Chr) %>%
  summarise(max_bp = max(position)) %>%
  mutate(offset = lag(cumsum(max_bp), default = 0))

# Step 2: Join offsets to main data
allsnp_manhattan <- all_snpsOrdered %>%
  left_join(scaffold_offsets, by = "Chr") %>%
  mutate(genomic_pos = position + offset)

#----------------------------------------------------------

#need to plot one at a time so that "Both" can be seen
ggplot() +
  # First: PC1_only (bottom layer, blue)
  geom_point(data = allsnp_manhattan %>% filter(SNP_status == "PC1_only"),
             aes(x = genomic_pos, y = logp),
             color = "blue", alpha = 0.7, size = 2) +

  # Second: PC2_only (middle layer, black)
  geom_point(data = allsnp_manhattan %>% filter(SNP_status == "PC2_only"),
             aes(x = genomic_pos, y = logp),
             color = "black", alpha = 0.7, size = 2) +

  # Third: Both (top layer, red)
  geom_point(data = allsnp_manhattan %>% filter(SNP_status == "Both"),
             aes(x = genomic_pos, y = logp),
             color = "red", alpha = 0.9, size = 3) +

  labs(x = "Genomic Position", y = "-log10(p-value)", color = "SNP Origin") +
  theme_bw()

#ggsave("~/projects/eco_genomics_2025/population_genomics/mydocs/SNP_Manhattan.png", plot = SNP_Manhattan, width = 6, height = 4, dpi = 300)
```
5. Signatures of Selection at the 
```{r}
#areas of selection (including overlap with Tajima's D and Pi)
# Ensure Chr is a factor with scaffold order
# Ensure Chr is a factor with scaffold order
allsnp_manhattan$Chr <- factor(allsnp_manhattan$Chr, levels = unique(allsnp_manhattan$Chr))

# Compute offsets for each scaffold
scaffold_offsets <- allsnp_manhattan %>%
  group_by(Chr) %>%
  summarise(max_bp = max(position), .groups = "drop") %>%
  mutate(offset = lag(cumsum(max_bp), default = 0))



ggplot() +
  geom_rect(data = selection_windows,
            aes(xmin = genomic_start, xmax = genomic_end, ymin = -Inf, ymax = Inf),
            fill = "violet", alpha = 0.3) +
  
  geom_point(data = allsnp_manhattan %>% filter(SNP_status == "PC1_only"),
             aes(x = genomic_pos, y = logp),
             color = "blue", size = 1, alpha = 0.7) +

  geom_point(data = allsnp_manhattan %>% filter(SNP_status == "PC2_only"),
             aes(x = genomic_pos, y = logp),
             color = "black", size = 1, alpha = 0.7) +

  geom_point(data = allsnp_manhattan %>% filter(SNP_status == "Both"),
             aes(x = genomic_pos, y = logp),
             color = "red", size = 1.5, alpha = 0.9) +

  labs(
    x = "Genomic Position",
    y = "-log10(p-value)",
    title = "PCA Outliers + Diversity-Based Selection Regions"
  ) +
  theme_bw()

```


