---
title: "HW1"
author: "SES"
date: "2025-09-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/projects/eco_genomics_2025/population_genomics")
```

1. Import our libraries for plotting 
```{r}
#load plotting libraries 
library(tidyverse)
library(ggplot2)
library(dplyr)
library(ggVennDiagram)
library(viridisLite)
library(ggpubr)
library(qqman)
```
2.Using Nucleotide diversity to find selection signatures
```{r} 
setwd("myresults/ANGSD/diversity")

theta <- read.table("_ALL_win50000_step50000.thetas", sep = "\t", header = T)

head(theta)
#Scaling Watterson's theta by nSites
theta$tWsite = theta$tW/theta$nSites 

#Scaling Pi by nSites
theta$tPsite = theta$tP/theta$nSites 

#check!
head(theta)
summary(theta)

#rearrange theta table 
theta2 = theta %>% 
  arrange(Chr, WinCenter) %>%
  filter(Chr > 100)

#compare pi and thetaW together and look for deviation 

#subsetting rows that are evidence of selective sweep
theta2[which(theta2$Tajima < -1.5 & theta2$tPsite < 0.001),]

summary(theta2)
sum(theta2$nSites)

```

3. Using PCA to find selection signatures 
```{r}
setwd("myresults/ANGSD/PCA_ADMIX")

#making the PCA with just the Red Spruce 

allRScov <- as.matrix(read.table("allRS_poly_K3.cov"))

allRSPCA <- eigen(allRScov)

var <- round(allRSPCA$values/sum(allRSPCA$values),3)

var[1:3]

#Bring in the bam list files and extract sample name information. Line 40 is the data frame with the list of names. The last number is the number of samples (1-95)
names <- read.table("allRS_bam.list")
names <- unlist(strsplit(basename(as.character(names[,1])), split = ".sorted.rmdup.bam"))
split = strsplit(names, "_")
pops <- data.frame(names[1:95], do.call(rbind, split[1:95]))
names(pops) = c("Ind", "Pop", "Row", "Col")

s <- read.table("allRS_poly_K3.selection")

# convert test statistic to p-value
pval_PC1 <- as.data.frame(1-pchisq(s[,1],1))
names(pval_PC1) = "p_PC1"

pval_PC2 <- as.data.frame(1-pchisq(s[,2],1))
names(pval_PC2) = "p_PC2"

#read in the sites: data on the snp positions themselves. Pairs the site information with sites that were selected for analysis in the outlier test.

PCAngsd_sites <- read.table("allRS_poly_K3.sites")
MAF_all_sites <- read.table("/gpfs1/cl/ecogen/pbio6800/PopulationGenomics/ANGSD/allRS_poly.mafs.gz", header=T)

pos_filtered <- cbind(MAF_all_sites,PCAngsd_sites)[which(PCAngsd_sites==1),]

dim(pos_filtered)

# Combine the locus info with the PC outlier p-values:

loci_pvals = cbind(pos_filtered, pval_PC1, pval_PC2)


# get all the outliers with p-values below some cutoff
cutoff=1e-3   # equals p<0.001

#table of selection outliers for each axis including the contig, snp position, and p_values 
outliers_PC1 <- loci_pvals %>%
                    filter(p_PC1<cutoff) %>%
                    select(chromo,position,p_PC1,p_PC2)
                    
outliers_PC2 <- loci_pvals %>%
                    filter(p_PC2<cutoff) %>%
                    select(chromo,position,p_PC1,p_PC2)


# how many outlier loci < the cutoff?
dim(outliers_PC1)[1]

dim(outliers_PC2)[1]

# how many unique contigs harbor outlier loci?
outlier_contigs_PC1 = unique(outliers_PC1$chromo)

length(outlier_contigs_PC1)

outlier_contigs_PC2 = unique(outliers_PC2$chromo)

length(outlier_contigs_PC2)
```
4. comparing significant values of the PCA outlier contigs with Tajima's D and Pi 
```{r}
#Checking colnames of tajima'sD and Pi (theta2 table) selection signatures
colnames(theta2)

#need to rename Chromo to Chr in the PCA outliers table to be able to compare the 3 tables 
#PC1
outliers_PC1 <- rename(outliers_PC1, Chr=chromo)
#PC2
outliers_PC2 <- rename(outliers_PC2, Chr=chromo)

##OVERLAP OF CONTIGS WITH SELECTION SIGNTURES (Venn Diagram)
venn_data <- list( 
  "TajimaD and Pi" = theta2$Chr, 
  "PC1 outliers" = outlier_contigs_PC1, 
  "PC2 outliers" = outlier_contigs_PC2)

ggVennDiagram(venn_data, label_alpha = 0.5, label="count") +
  scale_fill_viridis_c(option = "C") +
  theme(
    legend.position = "none",
    venn.setLabel = element_text(size = 12, color = "black"),
    plot.title = element_text(hjust = 0.5, size = 16, face = "bold")
  )
```

5. determine areas of selection across the genome 
```{r}
#SNPS with Selection signatures

#combining the PC1 and PC2 outliers tables, getting rid of overlaps  
all_snps <- unique(rbind(outliers_PC1, outliers_PC2))

# Create labels for plotting
all_snps$SNP_status <- ifelse((all_snps$p_PC1<0.001) & (all_snps$p_PC2<0.001), "Both",
                       ifelse(all_snps$p_PC1<0.001, "PC1_only",
                       ifelse(all_snps$p_PC2<0.001, "PC2_only", "None")))

# Calculate -log10(p-value) from either axis (choose one for y-axis)
all_snps$logp <- -log10(pmin(all_snps$p_PC1, all_snps$p_PC2, na.rm = TRUE))

#-------Making Chr and SNP positions into Genomic Positions-------------------------------

# 1. Make sure 'Chr' is in the correct order 
all_snps <- all_snps %>%
  mutate(Chr = factor(Chr, levels = unique(Chr)))

# 2. Calc max position for each chromosome
chrom_sizes <- all_snps %>%
  group_by(Chr) %>%
  summarise(max_pos = max(position)) %>%
  arrange(Chr)

# 3. Calculate cumulative lengths
chrom_sizes <- chrom_sizes %>%
  mutate(cum_len = cumsum(lag(max_pos, default = 0)))

# 4. Join cumulative lengths back to data
all_snps <- all_snps %>%
  left_join(chrom_sizes %>% select(Chr, cum_len), by = "Chr") %>%
  mutate(genomic_pos = position + cum_len)
#-----------------------------------------------------------------------------------------

#ordering the SNPs in "Both" last
all_snps <- all_snps %>%
  arrange(SNP_status != "Both")  # This puts "Both" rows at the end

ggplot(all_snps, aes(x = genomic_pos, y = logp, color = SNP_status)) +
  geom_point(alpha = 0.7) +
  scale_color_manual(values = c("PC1_only" = "blue", "PC2_only" = "black", "Both" = "red"),
                     labels = c("PC1_only" = "PC1 only", "PC2_only" = "PC2 only", "Both" = "Both")) +
  labs(x = "Genomic Position",
       y = "-log10(p-value)",
       color = "SNP Origin") +
  
  theme_bw()

```

```{r}
#areas of selection (including overlap with Tajima's D and Pi)


