---
title: "PCA_Admix"
author: "SES"
date: "2025-09-25"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir="/users/s/s/sstover/projects/eco_genomics_2025/population_genomics/myresults/ANGSD/PCA_ADMIX")

```

Load Libraries for plotting
```{r}
library(ggplot2)
library(ggpubr)
```
First let's estimate our PCA
```{r}
COV <- as.matrix(read.table("RSBS_poly_K2.cov")) # read in the genetic covariance matrix

PCA <- eigen(COV) # extract the principal components from the COV matrix
```
How much variance is explained by the first PCA/first eigenvalue?
```{r}
var <- round(PCA$values/sum(PCA$values),3)

var[1:3]
```
A screen plot of the eigenvalues of the PCA shows you each axis exlplains variance in the genotypes
```{r}
barplot(var, 
        xlab="Eigenvalues of the PCA", 
        ylab="Proportion of variance explained")
```
Bring in the bam.list file and extract the sample info to prepare for plotting:
```{r}
names <- read.table("RSBS_bam.list")
names <- unlist(strsplit(basename(as.character(names[,1])), split = ".sorted.rmdup.bam"))
split = strsplit(names, "_")
pops <- data.frame(names[1:113], do.call(rbind, split[1:113]))
names(pops) = c("Ind", "Pop", "Row", "Col")

data=as.data.frame(PCA$vectors)
data=data[,c(1:3)]
data= cbind(data, pops)
```
Now we can plot the PCA:
```{r}
ggscatter(data, x = "V1", y = "V2",
          color = "Pop",
          mean.point = TRUE,
          star.plot = TRUE,
          main="Red spruce-Black spruce genetic PCA") +
  theme_bw(base_size = 13, base_family = "Times") +
  labs(x = paste0("PC1: (",var[1]*100,"%)"), y = paste0("PC2: (",var[2]*100,"%)"))
```
Next, we can look at the admixture clustering at K=2

Import the ancestry scores (these are the .Q files)
```{r}
q <- read.table("RSBS_poly_K2.admix.2.Q", sep=" ", header=F)

K=dim(q)[2] #Find the level of K modeled

## rank order the samples according to population code
ord<-order(pops[,2])
```
Now we make the plot... 
```{r}
q <- read.table("RSBS_poly_K2.admix.2.Q", sep=" ", header=F)

K=dim(q)[2] #Find the level of K modeled

## rank order the samples according to population code
ord<-order(pops[,2])
```

